developer/src/kmc-kmn/src/compiler/embed-osk/embed-osk-kvk.ts

```diff
@@ -0,0 +1,158 @@
+import { KmnCompilerMessages } from "../kmn-compiler-messages.js";
+import { oskLayouts } from "./osk-layout.js";
+
+type VirtualKey = number;
+type LayerBag = Map<VirtualKey, KMXPlus.KeysKeys>;
+
+export class EmbedOskKvkInKmx {
+
+  constructor(private callbacks: CompilerCallbacks) {
+  }
+
+  /**
+   * Transform a .kvk file to the KMX+ format
+   */
+  public transformVisualKeyboardToKmxPlus(kmx: KMXPlus.KMXPlusFile, vk: VisualKeyboard.VisualKeyboard): void {
+
+    // TODO-EMBED-OSK-IN-KMX: if(displayMap) {
+    //   // Remap using the osk-char-use-rewriter
+    //   Osk.remapVisualKeyboard(vk, displayMap);
+    // }
+
+    const layerBags = this.buildLayerBags(vk, kmx.kmxplus.strs, kmx.kmxplus.keys);
+    const form = this.buildForm(vk, layerBags, kmx.kmxplus.strs);
+    kmx.kmxplus.layr.forms.push(form);
+
+    // For now, we only support dotted circle (U+25CC) as our base character
+    kmx.kmxplus.disp.baseCharacter = kmx.kmxplus.strs.allocString('\u25cc');
+  }
+
+  /**
+   * Build the layout of keys, with gaps for missing keys, for the given form --
+   * either ANSI (US) or ISO (EU), which are the only two supported layouts in
+   * .kvk
+   */
+  private buildForm(vk: VisualKeyboard.VisualKeyboard, layerBags: Map<number, LayerBag>, strs: KMXPlus.Strs) {
+    const baseLayoutName = 'en-us'; // This is the only value we support for 19.0
+    const formName: KMXPlus.LayrFormHardware =
+      vk.header.flags & VisualKeyboard.VisualKeyboardHeaderFlags.kvkh102
+      ? KMXPlus.LayrFormHardware.iso
+      : KMXPlus.LayrFormHardware.us;
+
+    const form = new KMXPlus.LayrForm();
+
+    form.baseLayout = strs.allocString(baseLayoutName);
+    form.flags = 0;
+
+    if(vk.header.flags & VisualKeyboard.VisualKeyboardHeaderFlags.kvkhDisplayUnderlying) {
+      form.flags |= KMXPlus.LayrFormFlags.showBaseLayout;
+    }
+
+    if(vk.header.flags & VisualKeyboard.VisualKeyboardHeaderFlags.kvkhAltGr) {
+      form.flags |= KMXPlus.LayrFormFlags.chiralSeparate;
+    }
+
+    // We will reserve space for the font facename to be rewritten, with a magic
+    // token that the package compiler will search for; see kmp-compiler.ts.
+    form.fontFaceName = strs.allocString(oskFontMagicToken);
+
+    // We only currently support 100% font size
+    form.fontSizePct = 100;
+    form.hardware = strs.allocString(formName);
+
+    // For hardware-style keyboards, device width is not relevant
+    form.minDeviceWidth = 0;
+
+    layerBags.forEach((keys, modifier) => {
+      const layr = new KMXPlus.LayrEntry();
+
+      // layr.id is not relevant for hardware keyboards, but we include it to
+      // make it easier to debug. We can use the existing KVK shift string
+      // generation, because we can only have KVK modifiers here, even though
+      // the LDML modifiers spec supports other modifiers
+      const vkShift = translateLdmlModifiersToVisualKeyboardShift(modifier);

```

Comment:
Why do we translate from LDML modifiers if we're working on a .kvk file (non-LDML)?